project/src/main/ui/chat/career-cutscene-library.gd

```diff
@@ -0,0 +1,309 @@
+		if preroll and not PlayerData.chat_history.is_chat_finished(preroll):
+			# don't exclude; preroll cutscene hasn't been played
+			continue
+		if postroll and not PlayerData.chat_history.is_chat_finished(postroll):
+			# don't exclude; postroll cutscene hasn't been played
+			continue
+		elif _preroll_tree.has(chat_key):
+			# calculate whether all child cutscenes have been played
+			var all_children_exhausted := true
+			for child in _preroll_tree.get(chat_key):
+				var child_key := _child_key(chat_key_roots, chat_key, child)
+				if not excluded_chat_keys.has(child_key):
+					all_children_exhausted = false
+			if not all_children_exhausted:
+				# don't exclude; child cutscenes haven't been played
+				continue
+		
+		# exclude; neither this node nor its children contains unplayed cutscenes
+		excluded_chat_keys[chat_key] = true
+	
+	return excluded_chat_keys
+
+
+## Returns a list of all chat keys in _preroll_tree in preroll order (leaves first.)
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 	cutscenes to choose between.
+##
+## Returns:
+## 	An array of String chat keys in postorder (leaves first)
+func _postorder_chat_keys(chat_key_roots: Array) -> Array:
+	var postorder_chat_keys := []
+	var chat_key_queue := chat_key_roots.duplicate()
+	while chat_key_queue:
+		var chat_key: String = chat_key_queue.pop_front()
+		
+		var children: Array = _preroll_tree.get(chat_key, [])
+		for child in children:
+			var child_key := _child_key(chat_key_roots, chat_key, child)
+			chat_key_queue.push_front(child_key)
+		
+		postorder_chat_keys.push_front(chat_key)
+	return postorder_chat_keys
+
+
+## Filters the list of potential chat key pairs, excluding the specified chat keys.
+##
+## The returned list includes the earliest numeric keys in each branch (0, 1, 2 are always played in order) as well as
+## ALL alphabetic keys in each branch (a, b, c are ordered arbitrarily)
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 		cutscenes to choose between.
+##
+## 	'exhausted_chat_keys': A set of chat keys for cutscenes the player has already seen. This includes the leaf
+## 		nodes for cutscenes which have been seen, and branch nodes where all child cutscenes have been seen.
+##
+## Returns:
+## 	A filtered list of dictionary entries. Each dictionary entry has a 'preroll' entry and/or a 'postroll' entry,
+## 	defining chat keys for cutscenes which play before or after a level. The list is filtered to exclude cutscenes
+## 	the player has either already seen, or should not see until later.
+func _filter_chat_key_pairs(chat_key_roots: Array, exhausted_chat_keys: Dictionary) -> Array:
+	var potential_chat_key_pairs := []
+	var chat_key_queue := chat_key_roots.duplicate()
+	var excluded_chat_keys := exhausted_chat_keys.duplicate()
```

Comment:
Add a variable comment describing how `excluded_chat_keys` differs from `exhausted_chat_keys`

Reply:
Refactored to remove `excluded_chat_keys` -- the tree is always traversed top-down so it's no longer necessary.

During development, the tree was traversed bidirectionally and we needed this field to prevent an infinite loop.